test/jdk/java/lang/ProcessBuilder/PipelineLeaksFD.java

```diff
@@ -0,0 +1,140 @@
+ * @run testng/othervm -DDEBUG PipelineLeaksFD
+ */
+
+@Test
+public class PipelineLeaksFD {
+    @DataProvider
+    public Object[][] builders() {
+        return new Object[][]{
+                {List.of(new ProcessBuilder("cat"))},
+                {List.of(new ProcessBuilder("cat"),
+                        new ProcessBuilder("cat"))},
+                {List.of(new ProcessBuilder("cat"),
+                        new ProcessBuilder("cat"),
+                        new ProcessBuilder("cat"),
+                        new ProcessBuilder("cat"),
+                        new ProcessBuilder("cat"))},
+        };
+    }
+
+    @Test(dataProvider = "builders")
+    void checkForLeaks(List<ProcessBuilder> builders) throws IOException {
+
+        Set<PipeRecord> pipesBefore = myPipes();
+        if (pipesBefore.size() < 3) {
+            System.out.println(pipesBefore);
+            Assert.fail("There should be at least 3 pipes before, (0, 1, 2)");
+        }
+
+        List<Process> processes = ProcessBuilder.startPipeline(builders);
+
+        OutputStream out = processes.get(0).getOutputStream();
+        out.write('a');
+        out.close();
+
+        Process last = processes.get(processes.size() - 1);
+        try (InputStream inputStream = last.getInputStream();
+             InputStream errorStream = last.getErrorStream()) {
+            byte[] bytes = inputStream.readAllBytes();
+            Assert.assertEquals(bytes.length, 1, "stdout bytes read");
+            byte[] errBytes = errorStream.readAllBytes();
+            Assert.assertEquals(errBytes.length, 0, "stderr bytes read");
+        }
+
+        processes.forEach(p -> waitForQuiet(p));
+
+        Set<PipeRecord> pipesAfter = myPipes();
+        printPipes(pipesAfter, "DEBUG: All Pipes After");
+        if (!pipesBefore.equals(pipesAfter)) {
+            Set<PipeRecord> missing = new HashSet<>(pipesBefore);
+            missing.removeAll(pipesAfter);
+            printPipes(missing, "Missing from pipesAfter");
+            Set<PipeRecord> extra = new HashSet<>(pipesAfter);
+            extra.removeAll(pipesBefore);
+            printPipes(extra, "Extra pipes in pipesAfter");
+            Assert.fail("More or fewer pipes than expected");
+        }
+    }
+
+    static void printPipes(Set<PipeRecord> pipes, String label) {
+        System.out.printf("%s: [%d]%n", label, pipes.size());
+        pipes.forEach(r -> System.out.printf("%-20s: %s%n", r.fd(), r.link()));
+    }
+
+    static void waitForQuiet(Process p) {
+        try {
+            int st = p.waitFor();
+            if (st != 0) {
+                System.out.println("non-zero exit status: " + p);
+            }
+        } catch (InterruptedException ie) {
+        }
+    }
+
+    /**
+     * Collect a Set of pairs of /proc fd paths and the symbol links that are pipes.
+     * @return A set of PipeRecords, possibly empty
+     * @throws IOException if reading the directory entries of "/proc/<pid>/fd/*" fails
```

Comment:
With these latest changes, it looks like the `IOException` will no longer be thrown by this method. Should we remove this from the javadoc then?